File: benchmarks/OCaml/safety/tacas2015/popl2007-1.ml
(* @from Program verification as probabilistic inference
         Sumit Gulwani and Nebojsa Jojic
         POPL 2007 *)

let rec f x y =
  if x < 100 then
    if x < 50 then
      f (x + 1) y
    else
      f (x + 1) (y + 1)
  else
    assert (y <= 100 && 100 <= y)
and main () = f 0 50

[@@@assert "typeof(main) <: unit -> unit"]

File: benchmarks/C/safety/popl2007-1.c
/* @from Program verification as probabilistic inference
         Sumit Gulwani and Nebojsa Jojic
         POPL 2007 */

extern error();

assume(int b) {
  if (!b) {
    assume(b);
  }
}

assert(int b) {
  if (!b) {
    error();
  }
}

main() {
  int x, y;
  x = 0;
  y = 50;
  while (x < 100) {
    if (x < 50) {
      x = x + 1;
    } else {
      x = x + 1;
      y = y + 1;
    }
  }
  assert(y == 100);
}
